docs/content/docs/ops/state/unaligned_checkpoints.md

```diff
@@ -0,0 +1,118 @@
+the current throughput as checkpoint barriers are effectively not embedded into the stream of data
+anymore.
+
+You should use unaligned checkpoints if your checkpointing durations are very high due to
+backpressure. Then, checkpointing time becomes mostly independent of the end-to-end latency. Be
+aware unaligned checkpointing adds to I/O to the state backends, so you shouldn't use it when the
+I/O to the state backend is actually the bottleneck during checkpointing.
+
+### Alignment timeout
+
+After enabling unaligned checkpoints, you can also specify the alignment timeout programmatically:
+
+```java
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+env.getCheckpointConfig().setAlignmentTimeout(Duration.ofSeconds(30));
+```
+
+or in the `flink-conf.yml` configuration file:
+
+```
+execution.checkpointing.alignment-timeout: 30 s
+```
+
+When activated, each checkpoint will still begin as an aligned checkpoint, but if the alignment time
+for some subtask exceeds this timeout, then the checkpoint will proceed as an unaligned checkpoint.
+
+## Limitations
+
+### Concurrent checkpoints
+
+Flink currently does not support concurrent unaligned checkpoints. However, due to the more
+predictable and shorter checkpointing times, concurrent checkpoints might not be needed at all.
+However, savepoints can also not happen concurrently to unaligned checkpoints, so they will take
+slightly longer.
+
+### Interplay with watermarks
+
+Unaligned checkpoints break with an implicit guarantee in respect to watermarks during recovery.
+Currently, Flink generates the watermark as the first step of recovery instead of storing the latest
+watermark in the operators to ease rescaling. In unaligned checkpoints, that means on recovery,
+**Flink generates watermarks after it restores in-flight data**. If your pipeline uses an
+**operator that applies the latest watermark on each record** will produce **different results** than for
+aligned checkpoints. If your operator depends on the latest watermark being always available, the
+workaround is to store the watermark in the operator state. In that case, watermarks should be
+stored per key group in a union state to support rescaling.
+
+### Certain connections are not checkpointed
+
+There are types of connections with properties that are impossible to keep with channel data stored
+in checkpoints. To preserve these characteristics and ensure no state corruption or unexpected
+behaviour, unaligned checkpoints are disabled for such connections. All other exchanges still
+perform unaligned checkpoints.
+
+**Pointwise connections**
+
+We currently do not have any hard guarantees on pointwise connections regarding data consistency.
+However, since data was structured implicitly in the same way as any preceding source or keyby, some
+users relied on this behaviour to divide compute-intensive tasks into smaller chunks while depending
+on orderliness guarantees.
```

Comment:
Could you elaborate a bit more what do you mean by "this behaviour"?

Reply:
this behaviour = `since data was structured implicitly in the same way as any preceding source or keyby`

Would characteristic be cleaner here?